src/CaptureFile/include/CaptureFile/CaptureFileOutputStream.h

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef CAPTURE_FILE_CAPTURE_FILE_OUTPUT_STREAM_H_
+#define CAPTURE_FILE_CAPTURE_FILE_OUTPUT_STREAM_H_
+
+#include <google/protobuf/message.h>
+
+#include <filesystem>
+#include <memory>
+
+#include "OrbitBase/Result.h"
+#include "capture.pb.h"
+
+namespace orbit_capture_file {
+
+// This class in used for creating new capture file from
+// a stream of ClientCaptureEvents. If the file already exists
+// it is going to be overwritten. Appending to the existing file
+// is not supported.
+//
+// Usage example:
+//
+// auto output_stream_or_error = CaptureOutputStream::Create("path/to/file.capture");
+// if (!has_value() {
+//   // Handle the error
+// }
+// auto& output_stream = output_stream_or_error.value();
+// for (...) {
+//   auto result = output_stream->WriteCaptureEvent(event);
+//   if (result.has_error() {
+//     // This is unrecoverable error the stream is closed and
+//     // the file is deleted.
+//     // Handle/Report the error.
+//     break;
+//   }
+// }
+//
+// output_stream->Close();
+//
+// Note: the stream will be closed on destruction if it was not explicitly closed before that.
+// Note: Write after close or error will result in CHECK failure.
+class CaptureFileOutputStream {
+ public:
+  virtual ~CaptureFileOutputStream() noexcept = default;
+  virtual ErrorMessageOr<void> WriteCaptureEvent(
+      const orbit_grpc_protos::ClientCaptureEvent& event) = 0;
+  virtual void Close() noexcept = 0;
```

Comment:
I'm not sure I understand the criterion you use for adding `noexcept` (considering we want to move away from exceptions altogether).

Reply:
this is a bit confusing, agree, it actually means "this does not allocate and therefore you shoudn't expect std::bad_alloc here"